File: scripts/populate-council-metrics-raw.py
#!/usr/bin/env python3
"""
scripts/populate-council-metrics-raw.py

Phase 2 (alternative): Aggregate VBA/BPC raw permit data (parsed XLSB CSV)
into council_metrics table per LGA per month.

Direct SQL UPSERT approach using Python + psycopg2.
Fall back to safe aggregate + print SQL statements for manual execution.

Run: python3 scripts/populate-council-metrics-raw.py
"""

import csv, os, sys, re
from collections import defaultdict

# ── Configuration ──
DATA_FILE = '/tmp/vba-data/2026-01_to_2026-03_raw.csv'

# ── LGA mapping ──
# (Built-in mapping: VBA name → lga_code)
# Synced from council_registry via Node (manual update if registry changes)
VBA_TO_LGA_CODE = {
    'Alpine': '20460', 'Ararat': '20260', 'Ballarat': '20570',
    'Banyule': '20660', 'Bass Coast': '20740', 'Baw Baw': '20830',
    'Bayside': '20910', 'Benalla': '21010', 'Boroondara': '21110',
    'Brimbank': '21180', 'Buloke': '21270', 'Campaspe': '21370',
    'Cardinia': '21430', 'Casey': '21470', 'Central Goldfields': '21610',
    'Colac-Otway': '21730', 'Corangamite': '21830', 'Darebin': '21890',
    'East Gippsland': '22030', 'Frankston': '22170', 'Gannawarra': '22200',
    'Glen Eira': '22310', 'Glenelg': '22370', 'Golden Plains': '22420',
    'Greater Bendigo': '22530', 'Greater Dandenong': '22670',
    'Greater Geelong': '22750', 'Greater Shepparton': '22860',
    'Hepburn': '23010', 'Hindmarsh': '23060', 'Hobsons Bay': '23110',
    'Horsham': '23230', 'Hume': '23270', 'Indigo': '23330',
    'Kingston': '23430', 'Knox': '23670', 'Latrobe': '23810',
    'Loddon': '23890', 'Macedon Ranges': '23970', 'Manningham': '24210',
    'Mansfield': '24250', 'Maribyrnong': '24330', 'Maroondah': '24410',
    'Melbourne': '24600', 'Melton': '24650', 'Merri-bek': '24700',
    'Mildura': '24790', 'Mitchell': '24860', 'Moira': '24920',
    'Monash': '24970', 'Moonee Valley': '25060', 'Moorabool': '25170',
    'Mornington': '25340', 'Mount Alexander': '25380', 'Moyne': '25450',
    'Mt Buller Alpine Resort': '90001',  # special purpose
    'Murrindindi': '25570', 'Nillumbik': '25610',
    'Northern Grampians': '25700', 'Port Philip': '25770',
    'Pyrenees': '25840', 'Queenscliff (B)': '25870',
    'South Gippsland': '26060', 'Southern Grampians': '26120',
    'Stonnington': '26210', 'Strathbogie': '26260', 'Surf Coast': '26310',
    'Swan Hill': '26350', 'Towong': '26450', 'Wangaratta': '26560',
    'Warrnambool': '26660', 'Wellington': '26700', 'West Wimmera': '26790',
    'Whitehorse': '26860', 'Whittlesea': '26910', 'Wodonga': '26960',
    'Wyndham': '27070', 'Yarra': '27110', 'Yarra Ranges': '27160',
    'Yarriambiack': '27210',
}


def parse_csv(path):
    """Parse CSV with proper quote handling."""
    rows = []
    with open(path, 'r', newline='') as f:
        reader = csv.reader(f)
        for row in reader:
            rows.append(row)
    return rows[0], rows[1:]  # header, data


def aggregate(header, rows):
    """Aggregate raw permit rows into per-LGA per-month stats."""
    col = {h: i for i, h in enumerate(header)}

    required = ['Site_Municipality', 'BASIS_Month_Y', 'BASIS_Month_M',
                'Reported_Cost_of_works', 'BASIS_Building_Use', 'BASIS_NOW']
    for c in required:
        assert c in col, f"Missing column: {c}"

    # Aggregation keys: (municipality, year, month)
    agg = defaultdict(lambda: {
        'total_count': 0, 'total_value': 0,
        'new_res_count': 0, 'new_res_value': 0,
        'multi_count': 0, 'multi_value': 0,
        'alt_count': 0, 'alt_value': 0,
        'comm_count': 0, 'comm_value': 0,
    })

    for row in rows:
        municip = row[col['Site_Municipality']].strip()
        if not municip:
            continue

        # Skip numeric-only values (postcode artifacts)
        if re.match(r'^\d+\.?\d*$', municip):
            continue

        year = int(float(row[col['BASIS_Month_Y']]))
        month = int(float(row[col['BASIS_Month_M']]))
        cost = float(row[col['Reported_Cost_of_works']] or 0)
        use = row[col['BASIS_Building_Use']].strip()
        now = float(row[col['BASIS_NOW']] or 0)

        key = (municip, year, month)
        r = agg[key]
        r['total_count'] += 1
        r['total_value'] += cost

        is_new = now == 1.0
        is_alt = now in (3.0, 4.0)
        is_commercial = use in ('Commercial', 'Retail', 'Industrial')

        if is_new and use == 'Domestic':
            r['new_res_count'] += 1
            r['new_res_value'] += cost
        elif is_new and use == 'Residential':
            r['multi_count'] += 1
            r['multi_value'] += cost
        elif is_alt:
            r['alt_count'] += 1
            r['alt_value'] += cost
        elif is_commercial:
            r['comm_count'] += 1
            r['comm_value'] += cost

    return agg


def main():
    print("[council-metrics-raw] Aggregating VBA raw permit data\n")

    if not os.path.exists(DATA_FILE):
        print(f"ERROR: Input CSV not found: {DATA_FILE}")
        sys.exit(1)

    header, rows = parse_csv(DATA_FILE)
    print(f"  Rows: {len(rows)}")
    print(f"  Cols: {len(header)}")

    agg = aggregate(header, rows)
    print(f"\n  Aggregated: {len(agg)} LGA-month records\n")

    # Build SQL UPSERT statements
    matched = 0
    unmatched = 0
    unmatched_names = set()
    sql_statements = []
    total_inserts = 0

    for (municip, year, month), rec in sorted(agg.items()):
        lga_code = VBA_TO_LGA_CODE.get(municip)

        if not lga_code:
            unmatched += 1
            unmatched_names.add(municip)
            continue

        matched += 1

        # Convert $ to $'000
        val_new_res = round(rec['new_res_value'] / 1000)
        val_multi = round(rec['multi_value'] / 1000)
        val_alt = round(rec['alt_value'] / 1000)
        val_comm = round(rec['comm_value'] / 1000)
        val_total = round(rec['total_value'] / 1000)
        avg_vpp = round(rec['total_value'] / rec['total_count'] / 1000) if rec['total_count'] > 0 else 0

        sql = f"""-- {municip} ({year}-{month:02d})
INSERT INTO council_metrics 
  (lga_code, report_year, report_month,
   permits_new_residential, permits_new_multi_unit, permits_alterations, permits_commercial, permits_total,
   value_new_residential, value_new_multi_unit, value_alterations, value_commercial, value_total,
   avg_value_per_permit, data_source)
VALUES (
  '{lga_code}', {year}, {month},
  {rec['new_res_count']}, {rec['multi_count']}, {rec['alt_count']}, {rec['comm_count']}, {rec['total_count']},
  {val_new_res}, {val_multi}, {val_alt}, {val_comm}, {val_total},
  {avg_vpp},
  'VBA/BPC Raw Permit Data'
)
ON CONFLICT (lga_code, report_year, report_month) DO UPDATE SET
  permits_new_residential = EXCLUDED.permits_new_residential,
  permits_new_multi_unit = EXCLUDED.permits_new_multi_unit,
  permits_alterations = EXCLUDED.permits_alterations,
  permits_commercial = EXCLUDED.permits_commercial,
  permits_total = EXCLUDED.permits_total,
  value_new_residential = EXCLUDED.value_new_residential,
  value_new_multi_unit = EXCLUDED.value_new_multi_unit,
  value_alterations = EXCLUDED.value_alterations,
  value_commercial = EXCLUDED.value_commercial,
  value_total = EXCLUDED.value_total,
  avg_value_per_permit = EXCLUDED.avg_value_per_permit,
  updated_at = NOW();"""

        sql_statements.append(sql)

    print(f"  Match results: {matched} matched, {unmatched} unmatched\n")

    if unmatched_names:
        print(f"  Unmatched ({len(unmatched_names)}):")
        for n in sorted(unmatched_names):
            print(f"    '{n}'")

    if unmatched > 0:
        print("\n  ❗ WARNING: Some municipalities could not be matched!")
        print("     Need to update VBA_TO_LGA_CODE mapping at top of script.")

    # Write SQL to file
    sql_output = '/tmp/vba-data/council_metrics_upsert.sql'
    os.makedirs('/tmp/vba-data', exist_ok=True)
    with open(sql_output, 'w') as f:
        f.write('-- VBA/BPC Raw Permit Data → council_metrics UPSERT\n')
        f.write('-- Generated: 2026-03 raw data (Jan-Mar 2026)\n')
        f.write(f'-- {matched} LGA-month records\n\n')
        f.write('\n'.join(sql_statements))

    print(f"\n  SQL written to: {sql_output}")

    # Also print summary
    print(f"\n  ── Summary by month ──")
    month_counts = defaultdict(int)
    for (m, y, mo), rec in sorted(agg.items()):
        if VBA_TO_LGA_CODE.get(m):
            month_counts[(y, mo)] += 1
    for (y, mo), cnt in sorted(month_counts.items()):
        print(f"    {y}-{mo:02d}: {cnt} LGAs")

    print(f"\n  ── Top 10 by permit count ──")
    by_count = sorted(
        [((m, y, mo), rec['total_count'], rec['total_value'])
         for (m, y, mo), rec in agg.items()
         if VBA_TO_LGA_CODE.get(m)],
        key=lambda x: -x[1]
    )[:10]
    for (m, y, mo), cnt, val in by_count:
        print(f"    {m:25s} {y}-{mo:02d}: {cnt:5d} permits  ${val/1e6:>8.1f}M")

    print(f"\n  Done! Run the SQL file, then REFRESH MATERIALIZED VIEW council_metrics_12m.")
    print(f"  SQL: psql $DATABASE_URL < {sql_output}")


if __name__ == '__main__':
    main()

File: scripts/populate-council-metrics-raw.mjs
#!/usr/bin/env node
/**
 * scripts/populate-council-metrics-raw.mjs
 *
 * Populate council_metrics from VBA/BPC raw permit data (XLSB → CSV export).
 *
 * Unlike populate-council-metrics.mjs (which expects pre-aggregated XLSX),
 * this script aggregates raw per-permit rows into per-LGA monthly stats.
 *
 * Input: CSV from pyxlsb-parsed XLSB raw data
 *   Columns used:
 *     - Site_Municipality (LGA name)
 *     - BASIS_Month_Y, BASIS_Month_M (year, month)
 *     - BASIS_Building_Use (Domestic, Commercial, Retail, etc.)
 *     - BASIS_NOW (Nature of Work: 1=New, 4=Alteration, etc.)
 *     - Reported_Cost_of_works (project cost $)
 *
 * Mapping:
 *   New Residential (houses)   = Building Use=Domestic + NOW=1 (New)
 *   New Multi Unit             = Building Use=Residential + NOW=1 (New)
 *   Alterations                = NOW=4 (Alteration) [all building uses]
 *   Commercial/Industrial      = Building Use=Commercial|Retail|Industrial [all NOW]
 *
 * Run: node scripts/populate-council-metrics-raw.mjs
 *
 * Safety: UPSERT only — never deletes
 */

import fs from 'fs';
import { neon } from '@neondatabase/serverless';
import dotenv from 'dotenv';

dotenv.config();

const sql = neon(process.env.DATABASE_URL, { fetchOptions: { cache: 'no-store' } });
const DATA_FILE = '/tmp/vba-data/2026-01_to_2026-03_raw.csv';

/**
 * Parse raw CSV (comma-separated, may contain quoted commas).
 * Returns { header: string[], rows: string[][] }
 */
function parseCsv(text) {
  const lines = [];
  let current = '';
  let inQuotes = false;
  
  for (let i = 0; i < text.length; i++) {
    const ch = text[i];
    if (ch === '"') {
      inQuotes = !inQuotes;
    } else if (ch === '\n' && !inQuotes) {
      lines.push(current);
      current = '';
    } else {
      current += ch;
    }
  }
  if (current.trim()) lines.push(current);
  
  const result = [];
  for (const line of lines) {
    const fields = [];
    let field = '';
    let q = false;
    for (let i = 0; i < line.length; i++) {
      const c = line[i];
      if (c === '"') {
        q = !q;
      } else if (c === ',' && !q) {
        fields.push(field.trim());
        field = '';
      } else {
        field += c;
      }
    }
    fields.push(field.trim());
    result.push(fields);
  }
  
  return { header: result[0], rows: result.slice(1) };
}

/**
 * Aggregate raw permit rows into per-LGA per-month stats.
 * Returns Map keyed by "municip|year|month" → aggregated record.
 */
function aggregatePermits(csv) {
  const idx = {};
  csv.header.forEach((h, i) => { idx[h.trim()] = i; });
  
  // Required columns
  const col = {
    municip: idx['Site_Municipality'],
    year: idx['BASIS_Month_Y'],
    month: idx['BASIS_Month_M'],
    cost: idx['Reported_Cost_of_works'],
    use: idx['BASIS_Building_Use'],
    now: idx['BASIS_NOW'],
  };
  
  const agg = new Map();
  
  for (const row of csv.rows) {
    const municip = row[col.municip]?.trim() || '';
    if (!municip) continue;
    
    const year = parseInt(row[col.year]) || 0;
    const month = parseInt(row[col.month]) || 0;
    const cost = parseFloat(row[col.cost]) || 0;
    const buildingUse = (row[col.use] || '').trim();
    const nowCode = parseFloat(row[col.now]) || 0;
    
    if (!year || !month) continue;
    
    const key = `${municip}|${year}|${month}`;
    if (!agg.has(key)) {
      agg.set(key, {
        lgaName: municip,
        year, month,
        totalCount: 0, totalValue: 0,
        newResCount: 0, newResValue: 0,
        multiCount: 0, multiValue: 0,
        altCount: 0, altValue: 0,
        commCount: 0, commValue: 0,
      });
    }
    
    const r = agg.get(key);
    r.totalCount++;
    r.totalValue += cost;
    
    // Classify by building use + nature of work
    const isNew = nowCode === 1;
    const isAlterationOrExtension = nowCode === 3 || nowCode === 4;
    const isCommercial = buildingUse === 'Commercial' || buildingUse === 'Retail' || buildingUse === 'Industrial';
    
    if (isNew && buildingUse === 'Domestic') {
      r.newResCount++;
      r.newResValue += cost;
    } else if (isNew && buildingUse === 'Residential') {
      r.multiCount++;
      r.multiValue += cost;
    } else if (isAlterationOrExtension) {
      r.altCount++;
      r.altValue += cost;
    } else if (isCommercial) {
      r.commCount++;
      r.commValue += cost;
    }
    // Other types (Hospital/Healthcare, Public Buildings) → not categorized separately
  }
  
  return agg;
}

async function loadLgaNameMap() {
  const rows = await sql`
    SELECT lga_code, lga_name, REGEXP_REPLACE(LOWER(lga_name), '[^a-z0-9]', '', 'g') AS slug
    FROM council_registry
  `;
  const nameMap = {};
  const slugMap = {};
  const fuzzyRows = [];
  for (const r of rows) {
    const nameLower = r.lga_name.toLowerCase().trim();
    nameMap[nameLower] = r.lga_code;
    // Also strip common suffixes
    const variants = [
      nameLower,
      nameLower.replace(/, (city|shire|rural city|borough|rural city|city of)$/i, ''),
      nameLower.replace(/^(city|shire|borough|rural city) of /i, ''),
    ];
    for (const v of [...new Set(variants)]) {
      const slug = v.replace(/[^a-z0-9]/g, '');
      slugMap[slug] = r.lga_code;
    }
    fuzzyRows.push(r);
  }
  return { nameMap, slugMap, rows: fuzzyRows };
}

function matchLga(vbaName, maps) {
  if (!vbaName) return null;
  const clean = vbaName.trim();
  
  // Direct match (case-insensitive)
  const lower = clean.toLowerCase();
  if (maps.nameMap[lower]) return maps.nameMap[lower];
  
  // Try with "City of" prefix/suffix variations
  const variants = [
    lower,
    `${lower} city council`,
    `${lower} council`,
    `city of ${lower}`,
    `shire of ${lower}`,
  ];
  for (const v of [...new Set(variants)]) {
    if (maps.nameMap[v]) return maps.nameMap[v];
  }
  
  // Slug match
  const slug = lower.replace(/[^a-z0-9]/g, '');
  if (maps.slugMap[slug]) return maps.slugMap[slug];
  
  // Fuzzy: partial match
  for (const r of maps.rows) {
    const rName = r.lga_name.toLowerCase();
    const rSlug = rName.replace(/[^a-z0-9]/g, '');
    // Check if VBA name is contained in registry name or vice versa
    if (rSlug.includes(slug) || slug.includes(rSlug)) {
      return r.lga_code;
    }
    // Check token overlap
    const vbaTokens = new Set(slug.split(/(?=[A-Z])/).filter(Boolean).map(t => t.toLowerCase()));
    const regTokens = new Set(rSlug.split(/(?=[A-Z])/).filter(Boolean).map(t => t.toLowerCase()));
    vbaTokens.forEach(t => {
      if (t === 'city' || t === 'council' || t === 'shire' || t === 'rural' || t === 'of') vbaTokens.delete(t);
    });
    regTokens.forEach(t => {
      if (t === 'city' || t === 'council' || t === 'shire' || t === 'rural' || t === 'of') regTokens.delete(t);
    });
    const intersection = [...vbaTokens].filter(t => regTokens.has(t));
    if (intersection.length >= Math.max(1, Math.min(vbaTokens.size, regTokens.size) - 1)) {
      return r.lga_code;
    }
  }
  
  return null;
}

async function main() {
  console.log('[council-metrics-raw] Aggregating VBA raw permit data → council_metrics\n');
  
  // Check input CSV
  if (!fs.existsSync(DATA_FILE)) {
    console.error(`[council-metrics-raw] ERROR: Input CSV not found: ${DATA_FILE}`);
    console.error('  First convert XLSB to CSV using pyxlsb');
    process.exit(1);
  }
  
  console.log(`  Input: ${DATA_FILE}`);
  const csvText = fs.readFileSync(DATA_FILE, 'utf8');
  const csv = parseCsv(csvText);
  console.log(`  Rows: ${csv.rows.length}`);
  console.log(`  Cols: ${csv.header.length}\n`);
  
  // Check council_registry
  const check = await sql`SELECT COUNT(*) FROM council_registry`;
  if (parseInt(check[0].count) === 0) {
    console.error('[council-metrics-raw] ERROR: council_registry is empty');
    process.exit(1);
  }
  console.log(`[council-metrics-raw] council_registry has ${check[0].count} LGAs\n`);
  
  // Load LGA mapping
  const maps = await loadLgaNameMap();
  
  // Aggregate
  const agg = aggregatePermits(csv);
  console.log(`[council-metrics-raw] Aggregated to ${agg.size} LGA-month records\n`);
  
  // Match and insert
  let matched = 0;
  let unmatched = 0;
  let inserted = 0;
  const unmatchedNames = new Set();
  
  // Process LGA-month pairs
  const entries = [...agg.entries()];
  for (const [key, rec] of entries) {
    const lgaCode = matchLga(rec.lgaName, maps);
    
    if (!lgaCode) {
      unmatched++;
      unmatchedNames.add(rec.lgaName);
      continue;
    }
    matched++;
    
    try {
      await sql`
        INSERT INTO council_metrics 
          (lga_code, report_year, report_month,
           permits_new_residential, permits_new_multi_unit, permits_alterations, permits_commercial, permits_total,
           value_new_residential, value_new_multi_unit, value_alterations, value_commercial, value_total,
           avg_value_per_permit, data_source)
        VALUES (
          ${lgaCode}, ${rec.year}, ${rec.month},
          ${rec.newResCount}, ${rec.multiCount}, ${rec.altCount}, ${rec.commCount}, ${rec.totalCount},
          ${Math.round(rec.newResValue / 1000)},  -- convert $ to $'000
          ${Math.round(rec.multiValue / 1000)},
          ${Math.round(rec.altValue / 1000)},
          ${Math.round(rec.commValue / 1000)},
          ${Math.round(rec.totalValue / 1000)},
          ${rec.totalCount > 0 ? Math.round(rec.totalValue / rec.totalCount / 1000) : 0},
          'VBA/BPC Raw Permit Data (Mar 2026)'
        )
        ON CONFLICT (lga_code, report_year, report_month) DO UPDATE SET
          permits_new_residential = EXCLUDED.permits_new_residential,
          permits_new_multi_unit = EXCLUDED.permits_new_multi_unit,
          permits_alterations = EXCLUDED.permits_alterations,
          permits_commercial = EXCLUDED.permits_commercial,
          permits_total = EXCLUDED.permits_total,
          value_new_residential = EXCLUDED.value_new_residential,
          value_new_multi_unit = EXCLUDED.value_new_multi_unit,
          value_alterations = EXCLUDED.value_alterations,
          value_commercial = EXCLUDED.value_commercial,
          value_total = EXCLUDED.value_total,
          avg_value_per_permit = EXCLUDED.avg_value_per_permit,
          updated_at = NOW()
      `;
      inserted++;
    } catch (err) {
      console.error(`    ✗ DB error for ${rec.lgaName} (rec.year}-${rec.month}): ${err.message}`);
    }
    
    // Show progress every 20
    if (inserted % 20 === 0) {
      process.stdout.write(`  ⟳ ${inserted} inserted...\r`);
    }
  }
  
  console.log(`\n  ✓ Match results: ${matched} matched, ${inserted} inserted, ${unmatched} unmatched`);
  
  if (unmatchedNames.size > 0) {
    console.log(`\n  Unmatched municipalities (${unmatchedNames.size}):`);
    for (const name of [...unmatchedNames].sort()) {
      console.log(`    "${name}"`);
    }
  }
  
  // Refresh materialized view
  console.log('\n  Refreshing council_metrics_12m...');
  try {
    await sql`REFRESH MATERIALIZED VIEW CONCURRENTLY council_metrics_12m`;
  } catch (e) {
    try {
      await sql`REFRESH MATERIALIZED VIEW council_metrics_12m`;
    } catch (e2) {
      console.warn(`  ! Cannot refresh: ${e2.message}`);
    }
  }
  console.log('  ✓ council_metrics_12m refreshed\n');
  
  // Summary
  const cov = await sql`
    SELECT report_year, report_month, COUNT(*) AS lgas
    FROM council_metrics
    GROUP BY report_year, report_month
    ORDER BY report_year, report_month
  `;
  
  const totalMetrics = await sql`SELECT COUNT(*) FROM council_metrics`;
  console.log('══════════════════════════════════════════');
  console.log('[council-metrics-raw] FINAL SUMMARY');
  console.log(`  Total council_metrics records: ${totalMetrics[0].count}`);
  console.log('  Periods covered:');
  for (const c of cov) {
    console.log(`    ${c.report_year}-${String(c.report_month).padStart(2, '0')}: ${c.lgas} LGAs`);
  }
  console.log('══════════════════════════════════════════\n');
  
  process.exit(0);
}

main().catch(err => {
  console.error('[council-metrics-raw] Fatal:', err);
  process.exit(1);
});
